Improve safety of `attempt` & `onErrorHandleWith`

oleg-py opened the pull request:
Closes #562.

I went ahead and made `attempt` safer too.

Incorporates tests from #559 comments and my latest work on `bracket`

### TODO:
- [x] Use `earlyStop` in guards

---

@alexandru should `earlyStop` be called if the evaluation of `tail` results on error? (I suppose it should).

I noticed some inconsitency that seems to not match my understanding of `earlyStop`:

```scala
private[tail] object IterantCompleteL {
  /**
    * Implementation for `Iterant#completeL`
    */
  final def apply[F[_], A](source: Iterant[F, A])
    (implicit F: Sync[F]): F[Unit] = {

    def loop(self: Iterant[F, A]): F[Unit] = {
      try self match { /* ...evicted... */ } catch {
        case ex if NonFatal(ex) =>
          source.earlyStop *> F.raiseError(ex)
      }
    }
```
Notice that it calls `source.earlyStop`. But in your comment for #559 you said:

> ... use a freaking var to hold the latest earlyStop reference 

and in this implementation the latest reference would be `self.earlyStop`. Most other operators name parameter inside loop `source` instead of `self`, shadowing the one that `apply` accepts, so they use latest `earlyStop`. Is that a bug?

Another question is the one I raised a number of times already 😄 .

```scala
Suspend(IO.raiseError(ex1), IO.raiseError(ex2)).onErrorHandleWith { ex =>
   /* is ex == ex1 or ex == e2 ? */
}
```

I assume per discussion in typelevel/cats-effect#113 we are going to swallow the one in earlyStop (which is akin to `release`), and there's nowhere it can go.

alexandru commented:
> Notice that it calls source.earlyStop. But in your comment for #559 you said:

That's not an inconsistency. It's a bug ☹️ as it should call `self.earlyStop`.

Unfortunately such bugs are easy when using inner functions and it's hard to test for it in this case.

alexandru commented:
About this:

```scala
Suspend(IO.raiseError(ex1), IO.raiseError(ex2)).onErrorHandleWith { ex =>
   /* is ex == ex1 or ex == e2 ? */
}
```

We can't swallow the one in `onErrorHandleWith` because it would probably violate the `MonadError` laws. `onErrorHandleWith` behaves like `flatMap` for errors. When the user returns an `F.raiseError` from `onErrorHandleWith`, we have to assume that's what he wants. And if an error gets thrown by the function given to `onErrorHandleWith`, we have to treat that as if the user returned an `F.raiseError`.

This is a different situation to `bracket`, as `onErrorHandleWith` is the equivalent of `catch`, not the equivalent of `finally`.

alexandru commented:
FYI — I am fixing `completeL`.

Avasil commented:
> FYI — I am fixing completeL.
Including entire #563?

oleg-py commented:
@alexandru I don't *return* raiseError 😛. As my Iterant, I have a `Suspend` of `rest` and `earlyStop` that both fail. Which error of two do I pass into `onErrorHandleWith`?

Fix for `completeL` will be nice - it should make `bracket` handle broken cursors

alexandru commented:
@Avasil not the entire issue, if you want you can take care of the others. Was thinking to put into practice my `var` idea to see if it works.

alexandru commented:
@oleg-py that's a tough call — prefer `ex1` (the error from `rest`), but do what's easy.

alexandru commented:
Created the PR - https://github.com/monix/monix/pull/567/files

Avasil commented:
Ok, I can take care of the others tomorrow, should be easy if your idea works, I'll do comparison benchmark too

alexandru commented on monix-tail/shared/src/main/scala/monix/tail/internal/IterantOnError.scala:
```diff
@@ -48,18 +48,23 @@ private[tail] object IterantOnError {
     }
 
+    def tailGuard(ffa: F[Iterant[F, A]]): F[Iterant[F, A]] =
+      ffa.handleError(f)
```
PS: gone ahead and opened a ticket for a `transformWith` in `MonadError`, fingers crossed — https://github.com/typelevel/cats/issues/2161

We can't wait for it though. It will be ages before that happens, if at all 😀

oleg-py replied:
That won't hurt. In the meantime, I think, users will appreciate correctness anyway 😄

oleg-py commented on monix-tail/shared/src/main/scala/monix/tail/internal/IterantOnError.scala:
```diff
@@ -101,16 +106,20 @@ private[tail] object IterantOnError {
 
   /** Implementation for `Iterant.attempt`. */
   def attempt[F[_], A](fa: Iterant[F, A])(implicit F: Sync[F]): Iterant[F, Either[Throwable, A]] = {
+    type ErrOrA = Either[Throwable, A]
```
👍

oleg-py commented on monix-tail/shared/src/test/scala/monix/tail/IterantOnErrorSuite.scala:
```diff
@@ -83,4 +85,64 @@ object IterantOnErrorSuite extends BaseTestSuite {
     }
   }
 
+  test("attempt & onErrorHandleWith should protect against broken continuations") { _ =>
```
I hope it's okay if I just extract common code to a method? 😏

oleg-py commented:
Ok, that should be it. I decided not to use the `var` trick - it seems to complicate the code with no benefit (as I cannot reduce the number of calls to `handleErrorWith`).

alexandru commented on monix-tail/shared/src/main/scala/monix/tail/internal/IterantOnError.scala:
```diff
@@ -48,18 +49,23 @@ private[tail] object IterantOnError {
       try fa match {
         case Next(a, lt, stop) =>
-          Next(a, lt.map(loop), stop)
+          Next(a, tailGuard(stop)(lt.map(loop)), stop)
```
Everything looks good except that I feel this call is overreaching. It should probably not include the `.map(loop)`:

```scala
tailGuard(lt, stop).map(loop)
```

For one this is because it's unnecessary and also because we're avoiding optimization opportunities — e.g. `lt` could be a `pure` value for which the `F[_]` implementation could bypass error handling.

alexandru commented:
@oleg-py is there some channel to reach you? Can we chat on Gitter? 😀

oleg-py commented:
I moved `map` outside the guard for `handleError`, but not for `attempt` since that way types don't align. Also, removed currying, it seems totally redundant now

alexandru commented:
Thanks @oleg-py
